pkg/wasm/httpfetcher.go

```diff
@@ -25,37 +27,50 @@ import (
 
 // HTTPFetcher fetches remote wasm module with HTTP get.
 type HTTPFetcher struct {
-	defaultClient  *http.Client
+	client         *http.Client
+	insecureClient *http.Client
 	initialBackoff time.Duration
 }
 
 // NewHTTPFetcher create a new HTTP remote wasm module fetcher.
-func NewHTTPFetcher() *HTTPFetcher {
+// requestTimeout is a timeout for each HTTP/HTTPS request.
+func NewHTTPFetcher(requestTimeout time.Duration) *HTTPFetcher {
+	if requestTimeout == 0 {
+		requestTimeout = 5 * time.Second
+	}
+	transport := http.DefaultTransport.(*http.Transport).Clone()
+	transport.TLSClientConfig = &tls.Config{InsecureSkipVerify: true}
 	return &HTTPFetcher{
-		defaultClient: &http.Client{
-			Timeout: 5 * time.Second,
+		client: &http.Client{
+			Timeout: requestTimeout,
+		},
+		insecureClient: &http.Client{
+			Timeout:   requestTimeout,
+			Transport: transport,
 		},
 		initialBackoff: time.Millisecond * 500,
 	}
 }
 
 // Fetch downloads a wasm module with HTTP get.
-func (f *HTTPFetcher) Fetch(url string, timeout time.Duration) ([]byte, error) {
-	c := f.defaultClient
-	if timeout != 0 {
-		c = &http.Client{
-			Timeout: timeout,
-		}
+func (f *HTTPFetcher) Fetch(ctx context.Context, url string, allowInsecure bool) ([]byte, error) {
+	c := f.client
+	if allowInsecure {
+		c = f.insecureClient
 	}
 	attempts := 0
-
 	b := backoff.NewExponentialBackOff()
 	b.InitialInterval = f.initialBackoff
 	b.Reset()
 	var lastError error
-	for attempts < 5 {
+	for attempts < DefaultWasmHTTPRequestMaxRetries {
 		attempts++
-		resp, err := c.Get(url)
+		req, err := http.NewRequestWithContext(ctx, http.MethodGet, url, nil)
```

Comment:
How does context timeout and requestTimeout correlate? if eiter is rtiggered we timeout?

Reply:
"context timeout" is triggered when the specified time is passed after calling `WithTimeout`,
and requestTimeout is a timeout for each http request attempt.

For example, if we have `context timeout = 7s, requestTimeout = 5s` and server does not respond, then, with very high possibility:
* First http request attempt might be timed out by `requestTimeout`. So, move to next attempts.
* Second attempt might be timed out by `context timeout`, so all the operation using this context will fail.

While checking your comment, I realized that the retry loop is still going after context timeout reach. Of course, each attempt will fail after the context timeout, but meaningless. I have patched it.